nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py

```diff
@@ -35,15 +79,68 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None, pretraine
-        self.register_buffer('inv_freq', inv_freq)
-        self.pretrained_max_position_embeddings = pretrained_max_position_embeddings
-        
+
+        if self.use_yarn:
+            self.seq_len_interpolation_factor = seq_len_interpolation_factor  # scale
+            self.max_positional_embeddings = max_positional_embeddings
+            self.extrapolation_factor = extrapolation_factor
+            self.attn_factor = attn_factor
+            self.beta_fast = beta_fast
+            self.beta_slow = beta_slow
+            self.dim = dim
+
+            self.yarn(self.seq_len_interpolation_factor)
+
+            self.max_seq_len_cached = self.max_positional_embeddings * self.seq_len_interpolation_factor
+            t = torch.arange(self.max_seq_len_cached, device=self.inv_freq.device, dtype=self.inv_freq.dtype)
+            freqs = torch.einsum("i , j -> i j", t, self.inv_freq)
+            # Different from paper, but it uses a different permutation in order to obtain the same calculation
+            emb = torch.cat((freqs, freqs), dim=-1) * self.mscale
+            self.register_buffer('emb', emb)
+
+        else:
+            inv_freq = 1.0 / (self.base ** (torch.arange(0, dim, 2).float() / dim))
+            self.register_buffer('inv_freq', inv_freq)
+
+    def yarn(self, scale):
+        pos_freqs = self.base ** (torch.arange(0, self.dim, 2).float() / self.dim)
+        inv_freq_extrapolation = 1.0 / pos_freqs
+        # inv_freq_interpolation = 1.0 / (self.seq_len_interpolation_factor * pos_freqs)
+        inv_freq_interpolation = 1.0 / (scale * pos_freqs)
+
+        low, high = find_correction_range(
+            self.beta_fast, self.beta_slow, self.dim, self.base, self.max_positional_embeddings
+        )
+        inv_freq_mask = (
+            1 - linear_ramp_mask(low, high, self.dim // 2).float()
+        ) * self.extrapolation_factor  # Get n-d rotational scaling corrected for extrapolation
+        inv_freq = inv_freq_interpolation * (1 - inv_freq_mask) + inv_freq_extrapolation * inv_freq_mask
+
+        self.register_buffer("inv_freq", inv_freq)
+        self.mscale = float(
+            get_mscale(scale) * self.attn_factor
+        )  # Get n-d magnitude scaling corrected for interpolation
+        global mscale
+        mscale = self.mscale
+
     def forward(self, max_seq_len, offset=0):
+        if self.use_yarn:
+            if max_seq_len > self.max_seq_len_cached:
+                self.max_seq_len_cached = max_seq_len
+                self.yarn(max_seq_len / self.max_positional_embeddings)
+                t = torch.arange(self.max_seq_len_cached, dtype=self.inv_freq.dtype).to(self.inv_freq.device)
+                freqs = torch.einsum("i , j -> i j", t, self.inv_freq)
+                # Different from paper, but it uses a different permutation in order to obtain the same calculation
+                emb = torch.cat((freqs, freqs), dim=-1).to(self.inv_freq.device) * self.mscale
```

Comment:
Do we need to remove self.mscale? although this doesn't affect continue training